File: autocontext/src/autocontext/security/child_process_env.py
"""Environment filtering for host-side child processes."""

from __future__ import annotations

import os
from collections.abc import Mapping

# Control-plane credentials authenticate the host process itself. Agent CLIs
# are less-trusted children and must not inherit material that lets them call
# back into the server as an operator.
CONTROL_PLANE_SECRET_ENV_KEYS = frozenset(
    {
        "AUTOCONTEXT_SERVER_AUTH_KEYS",
        "AUTOCONTEXT_SERVER_CREDENTIALS_FILE",
        "AUTOCONTEXT_SERVER_TOKEN",
        "AUTOCONTEXT_SERVER_TOKEN_FILE",
    }
)


def child_process_env_without_control_plane_secrets(
    source: Mapping[str, str] | None = None,
) -> dict[str, str]:
    """Copy an environment while removing host control-plane credentials."""

    return {
        key: value
        for key, value in (os.environ if source is None else source).items()
        if key.upper() not in CONTROL_PLANE_SECRET_ENV_KEYS
    }


def clear_control_plane_secrets_from_current_process() -> None:
    """Permanently consume server credentials before loading untrusted code."""

    for key in tuple(os.environ):
        if key.upper() in CONTROL_PLANE_SECRET_ENV_KEYS:
            del os.environ[key]

File: ts/src/security/child-process-env.ts
/** Host control-plane credentials must not cross into agent CLI children. */
export const CONTROL_PLANE_SECRET_ENV_KEYS = [
  "AUTOCONTEXT_SERVER_AUTH_KEYS",
  "AUTOCONTEXT_SERVER_CREDENTIALS_FILE",
  "AUTOCONTEXT_SERVER_TOKEN",
  "AUTOCONTEXT_SERVER_TOKEN_FILE",
] as const;

const CONTROL_PLANE_SECRET_ENV_KEY_SET = new Set<string>(CONTROL_PLANE_SECRET_ENV_KEYS);

export function isControlPlaneSecretEnvKey(key: string): boolean {
  return CONTROL_PLANE_SECRET_ENV_KEY_SET.has(key.toUpperCase());
}

export function childProcessEnvWithoutControlPlaneSecrets(
  source: NodeJS.ProcessEnv = process.env,
): NodeJS.ProcessEnv {
  const environment: NodeJS.ProcessEnv = {};
  for (const [key, value] of Object.entries(source)) {
    if (!isControlPlaneSecretEnvKey(key)) environment[key] = value;
  }
  return environment;
}

/**
 * Permanently consume control-plane credentials before loading or invoking
 * in-process, project-owned modules. Restoring them around async work would
 * create a process-wide race in concurrent servers.
 */
export function clearControlPlaneSecretsFromCurrentProcess(): void {
  for (const key of Object.keys(process.env)) {
    if (isControlPlaneSecretEnvKey(key)) delete process.env[key];
  }
}
